logger/glog/glog.go

```diff
@@ -1050,7 +1051,12 @@ func gzipFile(name string) error {
 	return os.Remove(name)
 }
 
+var rotationMutex sync.Mutex
+
 func (sb *syncBuffer) rotateOld(now time.Time) {
+	rotationMutex.Lock()
```

Comment:
Just to double check -- you don't think `rotateOld` could "pile up" if one scheduled rotation waits on the lock of another, and of another? I guess this would only happen if rotation were very slow and/or there were massive amounts of huge logs, but worth at least acknowledging?

A way around "pile up" might be to drop the scheduled rotation if the mutex is held, if you think it would be an issue

Reply:
They could "pile up", I considered this during development. I think it's not a problem, because the stacked executions should be cheap (mostly reading directory, iterating over entries). 
I was thinking about following alternatives:
1. As you said - rotations could be simply dropped. But that way we may break the log-size constraints for some time (until next rotation).
2. Limit the "piled up" queue to 1 goroutine. But this adds extra complexity - we need to notify the waiting goroutine, when active one ends execution, we need to ensure that only one goroutine waits for execution.
3. Use atomic to store information about running `rotateOld` and the need to run it again. Something like:
```
func rotateOld() {
    if atomic.CompareAndSwapUint32(atomic, 0, 1) {
        // execute rotation...
        if atomic.SwapUint32(atomic, 0) > 1 {
            go rotateOld()
        }
    } else {
        atomic.AddUint32(atomic, 1)
    }
```

3rd solution looks most correct, but IMHO it's too complicated. 'Piling up' should be rare and it's rather extreme case of wrong configuration or large logs.

Reply:
I agree -- I don't expect it to be an issue; gzip isn't that expensive, most user's won't have tons of logs initially (erm... like me ;), and it doesn't seem too bad to break a size constraint for a round or two of rotations (at worst). 

If it becomes an issue, you've got a good solve with `3` and we can throw that in if need be.

Reply:
Actual code for solution 3:
```
var rotationTime int64

func (sb *syncBuffer) rotateOld(now time.Time) {
	nanos := now.UnixNano()
	if atomic.CompareAndSwapInt64(&rotationTime, 0, nanos) {
		logs, err := getLogFiles()
		if err != nil {
			Fatal(err)
		}

		logs = excludeActive(logs)

		logs, err = removeOutdated(logs, now)
		if err != nil {
			Fatal(err)
		}

		logs, err = compressOrphans(logs)
		if err != nil {
			Fatal(err)
		}

		if MaxTotalSize > MaxSize {
			totalSize := getTotalSize(logs)
			for i := 0; i < len(logs) && totalSize > MaxTotalSize-MaxSize; i++ {
				err := os.Remove(filepath.Join(logs[i].dir, logs[i].name))
				if err != nil {
					Fatal(err)
				}
				totalSize -= logs[i].size
			}
		}

		if current := atomic.SwapInt64(&rotationTime, 0); current > nanos {
			go sb.rotateOld(time.Unix(0, current))
		}
	} else {
		atomic.StoreInt64(&rotationTime, nanos)
	}
}
```

It's actually not that bad (I was able to use only one atomic). Maybe we should push it now, without waiting for the issue? ;)

Reply:
If you like it, I like it. 👍

Reply:
That was exactly my intention.

Reply:
LGTM 👍